docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,420 @@
+     * the evidence.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] evidence_buffer The evidence buffer.
+     * @param[in] evidence_buffer_size The size of evidence_buffer in bytes.
+     * @param[in] collateral_buffer The collateral buffer.
+     * @param[in] collateral_buffer_size The size of collateral_buffer in bytes.
+     * @param[out] claims The list of claims.
+     * @param[out] claims_size The size of claims.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*verify_evidence)(
+        oe_attestaton_plugin_t* plugin_context,
+        const uint8_t* evidence_buffer,
+        size_t evidence_buffer_size,
+        const uint8_t* collateral_buffer,
+        size_t collateral_buffer_size,
+        uint_t** claims,
+        size_t* claims_size);
+};
+```
+
+Here is the rationale for each element in the plugin struct:
+
+- `format_id`
+  - Each plugin needs an unique identifier to distinguish itself.
+- `on_register` and `on_unregister`
+  - A plugin might require some setup or teardown when it is registered or
+    unregistered, so these functions are required. Furthermore, a plugin
+    might require configuration, which is why there is a `config_data`
+    parameter. The configuration data can be plugin specific, so no format is
+    specified in this proposal.
+- `get_evidence` and `free_evidence`
+  - Producing evidence is necessary for attestation.
+  - There is a `custom_claims` parameter because many attestation protocols
+    require the enclave to sign some claim from a relying party. For example,
+    many protocols follow the "challenge response" architecture, which requires
+    the enclave to sign a nonce from the relying party.
+- `get_collateral` and `free_collateral`
+  - Producing collateral is essential for attestation.
+  - Examples of collateral could be firmware measurements from the device's
+    manufacturer or CRLs from an X509 certificate.
+- `verify_evidence`
+  - Verifying evidence and collateral is essential for attestation.
+  - The `claims` field contains key-value pairs that can be verified by the
+    caller. This will have the similar contents as the `oe_identity_t` field
+    in the `oe_report_t` struct and any custom claims that were passed to the
+    `get_evidence` function.
+  - Should at minimum have the following claims (based of `oe_identity_t`):
+    - `id_version`: Version of the OE claims. Must be 0.
+    - `security_version`: Security version of the enclave. (ISVN for SGX).
+    - `attributes`: Values of the attributes flags for the enclave:
+      - `OE_REPORT_ATTRIBUTES_DEBUG`: The report is for a debug enclave.
+      - `OE_REPORT_ATTRIBUTES_REMOTE`: The report can be used for remote
+        attestation.
+    - `unique_id`: The unique ID for the enclave (MRENCLAVE for SGX).
+    - `signer_id`: The signer ID for the enclave (MRSIGNER for SGX).
+    - `product_id`: The product ID for the enclave (ISVPRODID for SGX).
+
+TODO:
```

Comment:
replace "TODO" with "Open issues"

Reply:
Do we want to start defining the header fields that will be needed by all plugins?

Reply:
The header was actually changed in the code sample. I'll add it to the doc.